### ISSUE TITLE
Add Text on video using LightOpenGlView

### ISSUE THREAD

**Issue by Rainbow87** [pedroSG94/rtmp-rtsp-stream-client-java#616]
Hi Pedro,
i'm using OpenGlView with TextObjectFilterRender to add text on top of video, it's working well.
However, because of my device is too slow, i want to use LightOpenGlView instead of OpenGlView  to render faster (fps of recording video must higher than 15).
I try to use TextObjectFilterRender  but it does not show text.
Pls help me how to add text?

Many thanks!

**Comment by pedroSG94** [comment]
You can't add text filter with LightOpenGlView because that view is created to work without filters and it provide a bit more performance.
I only can tell you set numFilter to 1 if you want get better performance (this is the limit of filters):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/app/src/main/res/layout/activity_open_gl.xml#L15

**Comment by Rainbow87** [comment]
Yes, i'm setting numFilter to 1 now, but the fps only reach about 9-12 fps, it's laggy. When use LightOpenGl can up to 15-18 fps.
I try to use LockCanvas and draw text on frame available callback, but lockcanvas failed

**Comment by pedroSG94** [comment]
That is normal. LightOpenGlView and OpenGlView is working with "layers".
In LightOpenGlVIew you have 1 layer (camera render) and in OpenGlView you have at least 3 layers (camera render + filter render * number of filters + screen render).

Of course you can join all layers in one if you always use same filter, but in my case, I did this more flexible (you can put any filter you want, any quantity and any order).

I recommend you join Text filter with LIghtOpenGlView render. Relevant classes:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/SimpleCameraRender.java
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/filters/object/TextObjectFilterRender.java
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/res/raw/camera_fragment.glsl
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/res/raw/object_fragment.glsl
Basically you need join both fragment and both render.

If you want try it with canvas (not recommend because your are using CPU and maybe you lose more fps). You can try add after this line:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/SimpleCameraRender.java#L120
Something similar to this (not tested):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/filters/AndroidViewFilterRender.java#L93

**Comment by Rainbow87** [comment]
Yes, i always use same one filter in my case, i need add text to recorded video only (dynamic text as time stamp...)
I try to implement text filter in ObjectFilterRender and in SimpleCameraRender but not successful,
I don't have experience in openGL Es
I will try follow as your guide
Thanks

**Comment by Rainbow87** [comment]
Hi Pedro,
i tried with lockCanvas and draw text but it's bad performance, many warning lock canvas failed, and lose fps. So this way is not good.
I also tried to join TextFilter with LightOpenGlView  by edit LightOpenGlView and SimpleCameraRender based on OpenGlView/ManagerRender/CameraRender/ScreenRender but OpenGL crashed in drawFrame().
Can you help me join all layers in one (camera render) with one text filter. I really don't know about openGL ES. I'm stuck! 

Thankful!

**Comment by pedroSG94** [comment]
Replace SimpleCameraRender in LighOpenGlView with this:
```java
package com.pedro.encoder.input.gl.render;

import android.content.Context;
import android.graphics.PointF;
import android.graphics.SurfaceTexture;
import android.graphics.Typeface;
import android.opengl.GLES11Ext;
import android.opengl.GLES20;
import android.opengl.Matrix;
import android.os.Build;
import androidx.annotation.RequiresApi;
import android.view.Surface;
import com.pedro.encoder.R;
import com.pedro.encoder.input.gl.Sprite;
import com.pedro.encoder.input.gl.TextureLoader;
import com.pedro.encoder.input.video.CameraHelper;
import com.pedro.encoder.utils.gl.GlUtil;
import com.pedro.encoder.utils.gl.SizeCalculator;
import com.pedro.encoder.utils.gl.StreamObjectBase;
import com.pedro.encoder.utils.gl.TextStreamObject;
import com.pedro.encoder.utils.gl.TranslateTo;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;

/**
 * Created by pedro on 21/02/18.
 */

@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
public class TextCameraRender {
  public final static String TAG = "SimpleCameraRender";

  private static final int FLOAT_SIZE_BYTES = 4;
  private static final int SQUARE_VERTEX_DATA_STRIDE_BYTES = 5 * FLOAT_SIZE_BYTES;
  private static final int SQUARE_VERTEX_DATA_POS_OFFSET = 0;
  private static final int SQUARE_VERTEX_DATA_UV_OFFSET = 3;

  private FloatBuffer squareVertex;

  private float[] MVPMatrix = new float[16];
  private float[] STMatrix = new float[16];
  private float[] rotationMatrix = new float[16];
  private float[] scaleMatrix = new float[16];

  private int[] texturesID = new int[1];

  private int program = -1;
  private int textureID = -1;
  private int uMVPMatrixHandle = -1;
  private int uSTMatrixHandle = -1;
  private int aPositionHandle = -1;
  private int aTextureCoordHandle = -1;

  private SurfaceTexture surfaceTexture;
  private Surface surface;
  private int streamWidth;
  private int streamHeight;
  private boolean isPortrait;

  //text item
  private int aTextureObjectHandle = -1;
  protected int uAlphaHandle = -1;
  private int uObjectHandle = -1;

  private FloatBuffer squareVertexObject;
  protected int[] streamObjectTextureId = new int[] { -1 };
  protected TextureLoader textureLoader = new TextureLoader();
  protected StreamObjectBase streamObject;
  private Sprite sprite;
  protected float alpha = 1f;
  protected boolean shouldLoad = false;

  public TextCameraRender() {
    Matrix.setIdentityM(MVPMatrix, 0);
    Matrix.setIdentityM(STMatrix, 0);
    float[] vertex = CameraHelper.getVerticesData();
    squareVertex = ByteBuffer.allocateDirect(vertex.length * FLOAT_SIZE_BYTES)
        .order(ByteOrder.nativeOrder())
        .asFloatBuffer();
    squareVertex.put(vertex).position(0);
    setRotation(0);
    setFlip(false, false);

    streamObject = new TextStreamObject();
    sprite = new Sprite();
    float[] vertices = sprite.getTransformedVertices();
    squareVertexObject = ByteBuffer.allocateDirect(vertices.length * FLOAT_SIZE_BYTES)
        .order(ByteOrder.nativeOrder())
        .asFloatBuffer();
    squareVertexObject.put(vertices).position(0);
  }

  public void setRotation(int rotation) {
    Matrix.setIdentityM(rotationMatrix, 0);
    Matrix.rotateM(rotationMatrix, 0, rotation, 0f, 0f, -1f);
    update();
  }

  public int getTextureId() {
    return textureID;
  }

  public SurfaceTexture getSurfaceTexture() {
    return surfaceTexture;
  }

  public Surface getSurface() {
    return surface;
  }

  public void updateFrame() {
    surfaceTexture.updateTexImage();
  }

  private void releaseTexture() {
    GLES20.glDeleteTextures(streamObjectTextureId.length, streamObjectTextureId, 0);
    streamObjectTextureId = new int[] { -1 };
  }

  public void setText(String text, float textSize, int textColor, Typeface typeface) {
    ((TextStreamObject) streamObject).load(text, textSize, textColor, typeface);
    shouldLoad = true;
  }

  public void drawFrame(int width, int height, boolean keepAspectRatio, int mode, int rotation,
      boolean isPreview, boolean flipStreamVertical, boolean flipStreamHorizontal) {
    if (shouldLoad) {
      releaseTexture();
      streamObjectTextureId = textureLoader.load(streamObject.getBitmaps());
      shouldLoad = false;
    }

    GlUtil.checkGlError("drawFrame start");
    surfaceTexture.getTransformMatrix(STMatrix);

    SizeCalculator.processMatrix(rotation, width, height, isPreview, isPortrait,
        flipStreamHorizontal, flipStreamVertical, mode, MVPMatrix);
    SizeCalculator.calculateViewPort(keepAspectRatio, mode, width, height, streamWidth,
        streamHeight);

    GLES20.glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
    GLES20.glClear(GLES20.GL_DEPTH_BUFFER_BIT | GLES20.GL_COLOR_BUFFER_BIT);

    GLES20.glUseProgram(program);

    squareVertex.position(SQUARE_VERTEX_DATA_POS_OFFSET);
    GLES20.glVertexAttribPointer(aPositionHandle, 3, GLES20.GL_FLOAT, false,
        SQUARE_VERTEX_DATA_STRIDE_BYTES, squareVertex);
    GLES20.glEnableVertexAttribArray(aPositionHandle);

    squareVertex.position(SQUARE_VERTEX_DATA_UV_OFFSET);
    GLES20.glVertexAttribPointer(aTextureCoordHandle, 2, GLES20.GL_FLOAT, false,
        SQUARE_VERTEX_DATA_STRIDE_BYTES, squareVertex);
    GLES20.glEnableVertexAttribArray(aTextureCoordHandle);

    squareVertexObject.position(SQUARE_VERTEX_DATA_POS_OFFSET);
    GLES20.glVertexAttribPointer(aTextureObjectHandle, 2, GLES20.GL_FLOAT, false,
        2 * FLOAT_SIZE_BYTES, squareVertexObject);
    GLES20.glEnableVertexAttribArray(aTextureObjectHandle);

    GLES20.glUniformMatrix4fv(uMVPMatrixHandle, 1, false, MVPMatrix, 0);
    GLES20.glUniformMatrix4fv(uSTMatrixHandle, 1, false, STMatrix, 0);
    //camera
    GLES20.glActiveTexture(GLES20.GL_TEXTURE0);
    GLES20.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, textureID);

    GLES20.glActiveTexture(GLES20.GL_TEXTURE1);
    GLES20.glUniform1i(uObjectHandle, 1);

    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, streamObjectTextureId[0]);
    //Set alpha. 0f if no image loaded.
    GLES20.glUniform1f(uAlphaHandle, streamObjectTextureId[0] == -1 ? 0f : alpha);
    //draw
    GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, 4);
    GlUtil.checkGlError("drawFrame end");
  }

  /**
   * Initializes GL state.  Call this after the EGL surface has been created and made current.
   */
  public void initGl(Context context, int streamWidth, int streamHeight) {
    isPortrait = CameraHelper.isPortrait(context);
    this.streamWidth = streamWidth;
    this.streamHeight = streamHeight;
    GlUtil.checkGlError("initGl start");
    String vertexShader = GlUtil.getStringFromRaw(context, R.raw.object_vertex);
    String fragmentShader = GlUtil.getStringFromRaw(context, R.raw.textcamera_fragment);

    program = GlUtil.createProgram(vertexShader, fragmentShader);
    aPositionHandle = GLES20.glGetAttribLocation(program, "aPosition");
    aTextureCoordHandle = GLES20.glGetAttribLocation(program, "aTextureCoord");
    uMVPMatrixHandle = GLES20.glGetUniformLocation(program, "uMVPMatrix");
    uSTMatrixHandle = GLES20.glGetUniformLocation(program, "uSTMatrix");
    aTextureObjectHandle = GLES20.glGetAttribLocation(program, "aTextureObjectCoord");
    uObjectHandle = GLES20.glGetUniformLocation(program, "uObject");
    uAlphaHandle = GLES20.glGetUniformLocation(program, "uAlpha");

    //camera texture
    GlUtil.createExternalTextures(1, texturesID, 0);
    textureID = texturesID[0];

    surfaceTexture = new SurfaceTexture(textureID);
    surfaceTexture.setDefaultBufferSize(streamWidth, streamHeight);
    surface = new Surface(surfaceTexture);
    GlUtil.checkGlError("initGl end");
  }

  public void release() {
    GLES20.glDeleteProgram(program);
    surfaceTexture = null;
    surface = null;
    releaseTexture();
    sprite.reset();
  }

  public void setFlip(boolean isFlipHorizontal, boolean isFlipVertical) {
    Matrix.setIdentityM(scaleMatrix, 0);
    Matrix.scaleM(scaleMatrix, 0, isFlipHorizontal ? -1f : 1f, isFlipVertical ? -1f : 1f, 1f);
    update();
  }

  private void update() {
    Matrix.setIdentityM(MVPMatrix, 0);
    Matrix.multiplyMM(MVPMatrix, 0, scaleMatrix, 0, MVPMatrix, 0);
    Matrix.multiplyMM(MVPMatrix, 0, rotationMatrix, 0, MVPMatrix, 0);
  }

  public void setAlpha(float alpha) {
    this.alpha = alpha;
  }

  public void setScale(float scaleX, float scaleY) {
    sprite.scale(scaleX, scaleY);
    squareVertexObject.put(sprite.getTransformedVertices()).position(0);
  }

  public void setPosition(float x, float y) {
    sprite.translate(x, y);
    squareVertexObject.put(sprite.getTransformedVertices()).position(0);
  }

  public void setPosition(TranslateTo positionTo) {
    sprite.translate(positionTo);
    squareVertexObject.put(sprite.getTransformedVertices()).position(0);
  }

  public PointF getScale() {
    return sprite.getScale();
  }

  public PointF getPosition() {
    return sprite.getTranslation();
  }

  public void setDefaultScale(int streamWidth, int streamHeight) {
    sprite.scale(streamObject.getWidth() * 100 / streamWidth,
        streamObject.getHeight() * 100 / streamHeight);
    squareVertexObject.put(sprite.getTransformedVertices()).position(0);
  }
}

```
textcamera_fragment.glsl:
```glsl
#extension GL_OES_EGL_image_external : require
precision mediump float;

uniform samplerExternalOES uSampler;
uniform sampler2D uObject;
uniform float uAlpha;

varying vec2 vTextureCoord;
varying vec2 vTextureObjectCoord;

void main() {
  vec4 samplerPixel = texture2D(uSampler, vTextureCoord);
  vec4 objectPixel = texture2D(uObject, vec2(1.0 - vTextureObjectCoord.y, 1.0 - vTextureObjectCoord.x));
  if (vTextureObjectCoord.x < 0.0 || vTextureObjectCoord.x > 1.0 ||
  vTextureObjectCoord.y < 0.0 || vTextureObjectCoord.y > 1.0) {
    gl_FragColor = samplerPixel;
  } else {
    gl_FragColor = mix(samplerPixel, objectPixel, objectPixel.a * uAlpha);
  }
}
```

Now add this method to LightOpenGlView and call it pressing any button to test:
```java
  public void setText() {
    simpleCameraRender.setText("hello world", 25f, Color.RED, null);
    simpleCameraRender.setScale(20, 20);
    simpleCameraRender.setPosition(TranslateTo.CENTER);
  }
```

**Comment by Rainbow87** [comment]
Seem that missing SizeCalculator class using in drawFrame(): 
SizeCalculator.processMatrix(rotation, width, height, isPreview, isPortrait,
        flipStreamHorizontal, flipStreamVertical, mode, MVPMatrix);
    SizeCalculator.calculateViewPort(keepAspectRatio, mode, width, height, streamWidth,
        streamHeight);

**Comment by pedroSG94** [comment]
That class is recent. Update library

**Comment by Rainbow87** [comment]
Follow your guide, now it can displays text, however text in vertical direction, i try to use canvas.rotate() in textAsBitmap() but it's not effect.
![2020-07-08_112121](https://user-images.githubusercontent.com/13791476/86875998-f8bb9500-c10d-11ea-9355-455f8ba9abe3.jpg)

**Comment by pedroSG94** [comment]
In textcamera_fragment.glsl replace:
```glsl
vec4 objectPixel = texture2D(uObject, vec2(1.0 - vTextureObjectCoord.y, 1.0 - vTextureObjectCoord.x));
```
to:
```glsl
vec4 objectPixel = texture2D(uObject, vec2(1.0 - vTextureObjectCoord.x, 1.0 - vTextureObjectCoord.y));
```

**Comment by Rainbow87** [comment]
Ok Pedro, i adjust in setScale and canvas rotate, it's effect and text displays in correct direction. The fps is good. Thank you very much!
